added sequence coverage map to gui results page

mgleeming opened the pull request:
Hi alphapept team,

Congratulations on a great project! I've been using alphapept over the last few days and I think this project has the rare combination of ease-of-use through the GUI and flexibility through well-documented scripts.

This could be a very useful project for the MS community and I'd like to support it if I can!

I saw your note welcoming contributions. I will admit that, despite some years as a python programmer, this is my first time contributing to the work of others. So, as a 'hello, world' offering, I added a simple sequence coverage map to the results page of the GUI.

Mostly, I wrote this to wrap my head around some of the code but maybe someone will find it useful. Please do feel free to use it or not as you like. If you've got comments on style or anything else, please let me know.

As extensions of this, I'm thinking it wouldn't be too difficult to add PSM, MS1 precursor and chromatogram viewers into the gui as well. Not sure what direction you want to take the project but I'd be happy to work on these among other things if there's interest.

Thanks again and all the best!
Michael

![AP](https://user-images.githubusercontent.com/16992583/132475915-dc0a3e51-38ee-407d-bf76-fc9c7efd63fe.png)

straussmaximilian reviewed (changes requested):
Hi Michael,

This is really awesome! I like your contribution a lot, and please more of this 👍👍👍. I made some comments on the code. My main suggestion would be if you could move the coverage calculations to the notebook and write a test for it? Then we would have the functional code within the continuous integration. If there are problems with the nbdev directly writing to the fasta.py should also be fine.

In terms of additional functions, we recently discovered the discussions function in Github https://github.com/MannLabs/alphapept/discussions and thought that this could be excellent to track what is interesting and what the community wants / and who is working on what.. At this point, we do prioritize on the algorithmic side, so any GUI integrations are very welcome and feel free to add what you find useful. (But also feel free to look at the algorithmic stuff in case.)

From the checks, there seems to be a bug with the CLA assistant, which I am going to fix in the meantime.

Best,

Max

mgleeming commented on alphapept/gui/results.py:
```diff
@@ -339,11 +341,130 @@ def scatter_plot(file: str, options: list):
+
+        with st.beta_expander("Sequence coverage map"):
+
+            protein_id = st.text_input(
```
Good idea - I'll do that!

mgleeming commented on alphapept/gui/results.py:
```diff
@@ -339,11 +341,130 @@ def scatter_plot(file: str, options: list):
+                for index, row in target_protein_peptide_matches.iterrows():
+                    peptide = row['naked_sequence']
+
+                    if peptide == '': # Found cases where naked_sequence was blank for all - Bug?
```
I noticed this when matching was activated. Without MBR, the naked sequences were correctly populated. I haven't followed it any further than that yet though. I'll confirm this with a separate dataset and open a new issue ticket if the result stands...

mgleeming commented on alphapept/gui/results.py:
```diff
@@ -339,11 +341,130 @@ def scatter_plot(file: str, options: list):
+
+                formatted_sequence = ''
+                counter = 0
+                for residue in residues:
```
Thanks :)

mgleeming commented on alphapept/gui/results.py:
```diff
@@ -339,11 +341,130 @@ def scatter_plot(file: str, options: list):
+                    matches = [m.start() for m in re.finditer('(?=%s)' %peptide, target_sequence)]
+                    for m in matches:
+                        for index in range(m, m+len(peptide)):
+                            assert residues[index]['res'] == target_sequence[index]
```
Yep, sorry, slipped through!

mgleeming commented:
Hi Max!

Thanks for your kind words and for the helpful comments! I'll make these changes tomorrow and update. 

That sounds like a great plan! I'm happy to focus on the GUI and build that out some more. I'll start a thread on the discussions page you mentioned about it so that we don't duplicate efforts with anyone else. 

Looking forward to working with you and the team!
Michael

PS: I'm quite a few timezones away in Australia (Melbourne) so we'll probably be trading night-time emails :)

mgleeming commented:
Hi Max,

I’ve incorporated your comments as above. To expand the GUI, I’d anticipate writing a bunch of miscellaneous functions to prepare data and, to avoid cluttering your existing files, I’ve added a new notebook ‘14_display.ipynb’ that can be a home for these. If you’re all focusing on the algorithmic side, this might help us to stay out of each other’s way.

Also, were you rebuilding the docs as separate commits? It looked that way from the ‘DOCUMENT’ tags on the commit messages. No problem – just wasn’t sure if I should do it or not. … I’ve left it for now.

All the best,
Michael

straussmaximilian commented on alphapept/_nbdev.py:
```diff
@@ -265,6 +265,7 @@
          "remove_mods": "13_export.ipynb",
          "ap_to_mq_sequence": "13_export.ipynb",
          "prepare_ap_results": "13_export.ipynb",
+         "calculate_sequence_coverage": "14_display.ipynb",
```
Good idea with the notebook and glad that the nbdev worked out.

straussmaximilian commented on alphapept/display.py:
```diff
@@ -0,0 +1,36 @@
+
+def calculate_sequence_coverage(target_sequence:str, peptide_list:list)->(int, int, float, list):
+    """
+    Calculate the percentage of a target protein covered by a list of peptides.
```
Thanks for sticking to the docstring convention.

straussmaximilian reviewed (approved):
Looks fantastic.

straussmaximilian commented:
Hi,
This looks all very good, and I can see that you mastered the nbdev structure. Great! 

I changed the base from master to develop as we typically run a complete performance check before merging to master. This will now go into v0.3.29 and we can all new things to the `changelog.md`.

Could you also update the branch to the latest develop? I hope this should fix the CLA assistant check. This was quite buggy now but should in principle ask for consent that we are allowed to merge your code contributions within the PR.

Some minor notes: 

You can see that some commits do match your profile avatar; it seems that you committed w/o setting the config for some commits.

<img width="82" alt="Screenshot 2021-09-11 at 00 17 20" src="https://user-images.githubusercontent.com/16521288/132923343-d3241d64-148f-4cd0-a714-22b2a7a62abf.png">

You can see this by typing `git shortlog -sne`.
<img width="417" alt="Screenshot 2021-09-11 at 00 18 27" src="https://user-images.githubusercontent.com/16521288/132923443-de0a1fbf-a36b-4264-a5f8-f65efed94cfa.png">

You could try to change this on your branch if you want but is typically cumbersome (https://help.github.com/en/github/using-git/changing-author-info).


We had some discussions today and will probably populate the projects/discussions page more in the coming days to make it a bit more transparent on what is being worked on.

As for the docs: Right now, the practice is to rebuild them in develop before they are merged to master. It is possible to test them beforehand locally but this also requires some fiddling.

Best,

Max

P.S.: Night emails are great. 😊

mgleeming commented:
Thanks Max,

Yep! The nbdev worked great! I'd never used it before but it seems like a great way to make notebooks more useful for development work rather than just data exploration!

I've merged the latest develop branch so hopefully that fixes the CLA problem - we'll see what happens :) Thanks for pointing out the config avatar thing. I'm working from 2 different machines which were configured a bit differently. I'll try to update them to match. 

Thanks for that -  I'll leave rebuilding the docs out of my commits and also start up some discussion threads for some GUI enhancements as well.

Have a nice weekend
Michael

straussmaximilian commented:
Hi Michael, the CLA assistant did not work for some reason.
As this is now causing more troubles than actual benefit at this point I will probably remove the action and decided to merge nonetheless.

Anyhow: Congratulations to the first contribution. I think this is awesome!

mgleeming commented:
Thanks Max! Happy to sign the CLA by some other means if it helps.
